### ISSUE TITLE
Question about flushWithCompletion

### ISSUE THREAD

**Issue by jcavar** [mixpanel/mixpanel-iphone#441]
In:

```
- (void)flushWithCompletion:(void (^)())handler
``` 
there is this code:

```
if (handler) {
    [self archive];
    dispatch_async(dispatch_get_main_queue(), handler);
}
```

I am wondering, what is the reason for archive being called only if handler is passed?

**Comment by samgreen** [comment]
I believe this is so we have more control internally as to when archive is called. For instance, when we enter the background, we want to `flush`, and call `archive` separately. As shown here: https://github.com/mixpanel/mixpanel-iphone/blob/master/Mixpanel/Mixpanel.m#L1232

Are you investigating some strange behavior? If you can give me some context, I'm sure I'll be able to help you get to the bottom of it.

**Comment by jcavar** [comment]
In our situation, we have very important event and we want to call flush immediately after it. There are two problems with this, first one is serial queue is not part of public API. The other one is if we call flush and there is no network connection, events won't be archived. We eventually found workarounds for both issues, but anyway, it is kind of weird that this:

```
mixPanel.flushWithCompletion {
}
 ```
does more then this:
 ```   
mixPanel.flushWithCompletion(nil)
```

**Comment by samgreen** [comment]
@jcavar I would be interested in hearing about the work around you used. I'm not comfortable exposing the serial queue as part of the public API, but I can understand the issues that you're having and I'd like to improve the API.

You also said: 
> The other one is if we call flush and there is no network connection, events won't be archived.

How did you come to this conclusion? Calling `flushWithCompletion:` will always archive, regardless of network activity.

**Comment by jcavar** [comment]
We dispatch archive and flush after some delay to workaround serial queue issue. 
Regarding ```flushWithCompletion:```, archive won't be called if you pass nil as handler, right?

**Comment by samgreen** [comment]
Sorry @jcavar to clarify what I mean, `archive` will only be called if you pass a completion block to `flushWithCompletion:` I assumed a block would be passed, otherwise you would just call `flush`. 

But again, network connectivity is unrelated to this behavior.

**Comment by jcavar** [comment]
Yes, you are right, what i meant is that if there is connectivity, we don't care weather archive is called or not because flush will complete. If there isn't connectivity and we call flush, it is possible that events aren't archived and we lose them.

This is why I explored implementation and then noticed that you call archive only if handler is not nil which was a bit weird.

**Comment by samgreen** [comment]
I think it's worthwhile for us to change this non-intuitive behavior.

**Comment by jcavar** [comment]
Great, thanks!

**Comment by samgreen** [comment]
This will be resolved in v2.9.5

**Comment by samgreen** [comment]
This is resolved in v2.9.5 which is now released.
